Enforce size bounds recursively on directory content

zabetak opened the pull request:
Enchance RequireFilesSize rule to be able to enforce size bounds in the entire content of a directory by optionally enabling the new recurscive flag.

The main motivation comes from the target directory that tends to store all kind of information from jar artifacts to test logs, reports, and many other things. It is not uncommon, that small seemingly innocent changes (e.g., adding a new test) can make the target directory unreasonably large consuming many GB of disk space. When such changes sneak into the project there is a high risk of causing CI/CD instability and intermittent failures that may be hard to notice.

Close #366

Following this checklist to help us incorporate your
contribution quickly and easily:

- [X] Your pull request should address just one issue, without pulling in other changes.
- [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [X] Each commit in the pull request should have a meaningful subject line and body.
  Note that commits might be squashed by a maintainer on merge.
- [X] Write unit tests that match behavioral changes, where the tests fail if the changes to the runtime are not applied.
  This may not always be possible but is a best-practice.
- [X] Run `mvn verify` to make sure basic checks pass.
  A more thorough check will be performed on your pull request automatically.
- [X] You have run the integration tests successfully (`mvn -Prun-its verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

slawekjaranowski commented on enforcer-rules/src/test/java/org/apache/maven/enforcer/rules/files/TestRequireFilesSize.java:
```diff
@@ -180,6 +183,41 @@ void testRequireFilesSizeSatisfyAny() throws EnforcerRuleException, IOException
 
+    @Test
+    void testDirectoryContentOverUpperBound() throws IOException {
+        Path d = Files.createTempDirectory("junit");
```
please use `temporaryFolder`

slawekjaranowski reviewed (changes requested):
Please also update documentation in `enforcer-rules/src/site/apt/requireFilesSize.apt.vm`

zabetak commented:
Many thanks for the review @slawekjaranowski ! I addressed your comments and updated the documentation in https://github.com/apache/maven-enforcer/pull/368/commits/6b56ba1bc6e4e831cd887ff5d9efd991c09c8cc9. I run `mvn site` and checked that `enforcer-rules/target/site/requireFilesSize.html` reflects the changes.

slawekjaranowski commented:
@zabetak thanks
